common/testing/testlogger/testlogger.go

```diff
@@ -0,0 +1,501 @@
+	failOnError  atomic.Bool
+	t            TestingT
+	mu           struct {
+		sync.RWMutex
+		expectations   map[Level]*list.List // Map[Level]List[matcher]
+		closed         bool
+		unexpectedErrs bool
+	}
+	logExpectations bool
+	logCaller       bool
+	level           zapcore.Level
+}
+
+// TestLogger is a log.Logger implementation that logs to the test's logger
+// _but_ will fail the test if log levels _above_ Warn are present
+type TestLogger struct {
+	wrapped log.Logger
+	state   *SharedTestLoggerState
+	tags    []tag.Tag
+}
+
+type LoggerOption func(*TestLogger)
+
+func DontFailOnError(t *TestLogger) *TestLogger {
+	t.state.failOnError.Store(false)
+	return t
+}
+
+func DontFailOnDPanic(t *TestLogger) *TestLogger {
+	t.state.failOnDPanic.Store(false)
+	return t
+}
+
+func LogExpectations(t *TestLogger) *TestLogger {
+	t.state.logExpectations = true
+	return t
+}
+
+// WithLogTags adds tags to the logs emitted by the underlying logger
+func WithLogTags(tags ...tag.Tag) LoggerOption {
+	return func(t *TestLogger) {
+		t.tags = tags
+	}
+}
+
+func WrapLogger(l log.Logger) LoggerOption {
+	return func(t *TestLogger) {
+		t.wrapped = l
+	}
+}
+
+func WithoutCaller() LoggerOption {
+	return func(t *TestLogger) {
+		t.state.logCaller = false
+	}
+}
+
+// SetLogLevel overrides the temporal test log level during this test.
+func SetLogLevel(tt CleanupCapableT, level zapcore.Level) LoggerOption {
+	return func(t *TestLogger) {
+		oldLevel := os.Getenv("TEMPORAL_TEST_LOG_LEVEL")
+		_ = os.Setenv("TEMPORAL_TEST_LOG_LEVEL", level.String())
+		tt.Cleanup(func() {
+			_ = os.Setenv("TEMPORAL_TEST_LOG_LEVEL", oldLevel)
+		})
+
+		t.state.level = level
+	}
+}
+
+var _ log.Logger = (*TestLogger)(nil)
+
+// NewTestLogger creates a new TestLogger that logs to the provided testing.T.
+// By default it will fail the test if Error, or DPanic is called.
+func NewTestLogger(t TestingT, opts ...LoggerOption) *TestLogger {
+	tl := &TestLogger{
+		state: &SharedTestLoggerState{
+			t:               t,
+			logExpectations: false,
+			level:           zapcore.DebugLevel,
+			logCaller:       true,
+		},
+	}
+	tl.state.failOnError.Store(true)
+	tl.state.failOnDPanic.Store(true)
+	tl.state.mu.expectations = map[Level]*list.List{
+		Debug: list.New(),
+		Info:  list.New(),
+		Warn:  list.New(),
+		Error: list.New(),
+	}
+	for _, opt := range opts {
+		opt(tl)
+	}
+	if tl.wrapped == nil {
+		ztl := zaptest.NewLogger(t,
+			zaptest.Level(tl.state.level),
+			zaptest.WrapOptions(
+				zap.AddStacktrace(zap.ErrorLevel),
+				zap.WithCaller(tl.state.logCaller),
+			),
+		)
+		tl.wrapped = log.NewZapLogger(ztl).Skip(1)
+	}
+
+	// Only possible with a *testing.T until *rapid.T supports `Cleanup`
+	if ct, ok := t.(CleanupCapableT); ok {
+		// NOTE(tim): We don't care about anything logged after the test completes. Sure, this is racy,
+		// but it reduces the likelihood that we see stupid errors due to testing.T.Logf race conditions...
+		ct.Cleanup(tl.Close)
+	}
```

Comment:
Note to OSS as new users of this code: if you call `t.Logf` after the test completes it will panic. When using this test logger you _must_ make sure to clean up your dependencies/goroutines before the test completes. The zaptest.Logger we rely upon uses `t.Log*` internally, so any logs that occur after the `testing.T` is marked as complete will cause a failure.

Ask me how I know 😅